mmedit/datasets/pipelines/augmentation.py

```diff
@@ -145,6 +148,190 @@ def __repr__(self):
+                top (int): Vertical coordinate of the top left
+                           corner of the crop box.
+                left (int): Horizontal coordinate of the top left
+                            corner of the crop box.
+                crop_h (int): Height of the crop box.
+                crop_w (int): Width of the crop box.
+        """
+        data_h, data_w = data.shape[:2]
+        area = data_h * data_w
+
+        for _ in range(10):
+            target_area = random.uniform(*scale) * area
+            log_ratio = (math.log(ratio[0]), math.log(ratio[1]))
+            aspect_ratio = math.exp(random.uniform(*log_ratio))
+
+            crop_w = int(round(math.sqrt(target_area * aspect_ratio)))
+            crop_h = int(round(math.sqrt(target_area / aspect_ratio)))
+
+            if 0 < crop_w <= data_w and 0 < crop_h <= data_h:
+                top = random.randint(0, data_h - crop_h)
+                left = random.randint(0, data_w - crop_w)
+                return top, left, crop_h, crop_w
+
+        # Fallback to central crop
+        in_ratio = float(data_w) / float(data_h)
+        if (in_ratio < min(ratio)):
+            crop_w = data_w
+            crop_h = int(round(crop_w / min(ratio)))
+        elif (in_ratio > max(ratio)):
+            crop_h = data_h
+            crop_w = int(round(crop_h * max(ratio)))
+        else:  # whole image
+            crop_w = data_w
+            crop_h = data_h
+        top = (data_h - crop_h) // 2
+        left = (data_w - crop_w) // 2
+        return top, left, crop_h, crop_w
+
+    def _crop(self, data, top, left, height, width):
+        data = data[top:top + height, left:left + width, ...]
+        return data
+
+    def __call__(self, results):
+        """Call function.
+
+        Args:
+            results (dict): A dict containing the necessary information and
+                data for augmentation.
+
+        Returns:
+            dict: A dict containing the processed data and information.
+        """
+        for k in self.keys:
+            top, left, crop_h, crop_w = self.get_params(
+                results[k], self.scale, self.ratio)
+            results[k] = self._crop(results[k], top, left, crop_h, crop_w)
+            results[k], _, _ = mmcv.imresize(
+                results[k],
+                self.crop_size,
+                return_scale=True,
+                interpolation=self.interpolation)
+        return results
+
+    def __repr__(self):
+        repr_str = self.__class__.__name__
+        repr_str += (f'(keys={self.keys}, crop_size={self.crop_size}, ')
+        return repr_str
+
+
+@PIPELINES.register_module()
+class RandomRotation:
+    """Rotate the image by angle.
+
+    Args:
+        keys (list[str]): The images to be rotated.
+        degrees (sequence or number): Range of degrees to select from.
+            If degrees is a number instead of sequence like (min, max),
+            the range of degrees will be (-degrees, +degrees).
```

Comment:
degrees (tuple(float) | float): If it is a tuple, it represents a range (min, max). If it is a float, the range is constructed as (-degrees, degrees).